### DISCUSSION TITLE
Performance overhead of subscriptions

### DISCUSSION THREAD

**Discussion by peppy** [realm/realm-dotnet#2775] [Q&A] [answered]
As we begin to introduce more usage of `SubscribeForNotifications` into our game, we are noticing `Realm.Refresh` calls are getting slower and slower. This seems dependent on the number of subscriptions active.

Calling `Realm.Refresh` from the main thread is basically a requirement of using realm, so we have the expectation that the overhead of calling it is low enough to not impede performance (unless obviously expensive operations were queued up since the last run). What we're seeing is a large overhead from a single realm update.

- To cause the overhead, the `SubscribeForNotifications` needs to be on a model being updated
- Threading doesn't matter - the update can be done on the main or an async thread and the same overhead is observed

Even when there's a lot of objects in a realm, my expectation is that since realm is aware of all ongoing operations, reporting said operations to a subscription should be of negligible cost, else it renders subscriptions useless for a large number of use cases.

## Real world implications

With no subscriptions, this is how deleting items from realm looks. `Realm.Refresh` calls do have an overhead, but it is low enough that I think we can potentially deal with it (0-5ms). As an aside, the rate of deletions here is only limited by the fact we are deleting over item per transaction, I am aware we can greatly improve this but it is out of scope of this report:

https://user-images.githubusercontent.com/191335/150315108-7d2f367f-731f-4e59-9019-47535b1d165e.mp4

To show the overhead of a single deletion operation, I added a sleep between deletions (else the game basically comes to a halt). With one subscription and deleting a single object from our realm, we see spikes of around 500ms:

https://user-images.githubusercontent.com/191335/150314528-868b9b2e-3d90-4b49-a04e-e87239552943.mov

With five subscriptions (some are only subscribed on the song select screen, which explains the increased overhead compared to the menu above):

https://user-images.githubusercontent.com/191335/150314565-39a4109f-35da-4d2e-b2ac-2a6134796f06.mov

The overhead looks to be native, as checked on macOS, then windows to confirm:

![unknown](https://user-images.githubusercontent.com/191335/150314765-fb26a4f5-6f9d-4f25-9bbd-db15dfbf3d21.png)

![unknown-2](https://user-images.githubusercontent.com/191335/150314802-c38d1eb4-8b2e-45eb-a1f6-99b3dbedacc8.png)

From the profiling output, I'd say this is likely a `realm-core` issue, but posting it here seems amicable to begin further investigation as usage and examples is in dotnet projects.

## Reproduction

I've written a barebones repro of this taking away the complexities of our project:

```csharp
using System;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Threading;
using NUnit.Framework;
using Realms;

namespace RealmPerformanceTest
{
    [TestFixture]
    public class RealmScratch
    {
        private readonly Stopwatch timer = new Stopwatch();

        [TestCase(1000)]
        [TestCase(10000)]
        [TestCase(100000)]
        [TestCase(1000000)]
        public void TestNestedContexts(int count)
        {
            ManualResetEventSlim gotInitialCallback = new ManualResetEventSlim();

            timer.Restart();

            string realmFilename = Path.Combine(Environment.CurrentDirectory, "test.realm");

            File.Delete(realmFilename);

            var configuration = new RealmConfiguration(realmFilename);

            using (var realm = Realm.GetInstance(configuration))
            {
                using (var transaction = realm.BeginWrite())
                {
                    for (int i = 0; i < count; i++)
                        realm.Add(new PersistedObject());

                    transaction.Commit();
                }
            }

            Console.WriteLine($"Running with realm size {new FileInfo(realmFilename).Length:g} ({count} objects)");

            using (var realm = Realm.GetInstance(configuration))
            {
                realm
                    .All<PersistedObject>()
                    .SubscribeForNotifications(NotifyCallback);

                while (!gotInitialCallback.IsSet)
                    TimedOp(realm.Refresh, "refresh");

                for (int i = 0; i < 10; i++)
                {
                    using (var transaction = realm.BeginWrite())
                    {
                        // make an arbitrary change to one object
                        realm.All<PersistedObject>().First(s => !s.ArbitraryBool).ArbitraryBool = true;
                        TimedOp(transaction.Commit, "commit");
                    }

                    TimedOp(realm.Refresh, "refresh");
                }
            }


            void NotifyCallback(IRealmCollection<PersistedObject> sender, ChangeSet changes, Exception error)
            {
                if (changes == null)
                    gotInitialCallback.Set();
            }
        }

        private void TimedOp(Func<bool> op, string name)
        {
            long before = timer.ElapsedMilliseconds;
            op();
            Console.WriteLine($"{name} took {timer.ElapsedMilliseconds - before} ms");
        }

        private void TimedOp(Action op, string name)
        {
            long before = timer.ElapsedMilliseconds;
            op();
            Console.WriteLine($"{name} took {timer.ElapsedMilliseconds - before} ms");
        }

        public class PersistedObject : RealmObject
        {
            public bool ArbitraryBool { get; set; }
        }
    }
}
```

Results for these runs are as follows:

```csharp
Running with realm size 8192 (1000 objects)
refresh took 3 ms
commit took 0 ms
refresh took 6 ms
commit took 0 ms
refresh took 0 ms
commit took 1 ms
refresh took 0 ms
commit took 0 ms
refresh took 0 ms
commit took 0 ms
refresh took 0 ms
commit took 0 ms
refresh took 1 ms
commit took 0 ms
refresh took 0 ms
commit took 0 ms
refresh took 1 ms
commit took 0 ms
refresh took 0 ms
commit took 1 ms
refresh took 0 ms
```

```csharp
Running with realm size 53248 (10000 objects)
refresh took 1 ms
commit took 0 ms
refresh took 2 ms
commit took 1 ms
refresh took 3 ms
commit took 0 ms
refresh took 3 ms
commit took 0 ms
refresh took 3 ms
commit took 0 ms
refresh took 2 ms
commit took 0 ms
refresh took 3 ms
commit took 0 ms
refresh took 2 ms
commit took 0 ms
refresh took 2 ms
commit took 0 ms
refresh took 2 ms
commit took 0 ms
refresh took 2 ms
```

```csharp
Running with realm size 512000 (100000 objects)
refresh took 15 ms
commit took 0 ms
refresh took 21 ms
commit took 22 ms
refresh took 21 ms
commit took 0 ms
refresh took 22 ms
commit took 1 ms
refresh took 22 ms
commit took 0 ms
refresh took 19 ms
commit took 1 ms
refresh took 19 ms
commit took 0 ms
refresh took 25 ms
commit took 0 ms
refresh took 21 ms
commit took 1 ms
refresh took 22 ms
commit took 1 ms
refresh took 21 ms
```

```csharp
Running with realm size 5603328 (1000000 objects)
refresh took 154 ms
commit took 1 ms
refresh took 207 ms
commit took 1 ms
refresh took 194 ms
commit took 1 ms
refresh took 192 ms
commit took 1 ms
refresh took 186 ms
commit took 2 ms
refresh took 196 ms
commit took 1 ms
refresh took 181 ms
commit took 1 ms
refresh took 187 ms
commit took 1 ms
refresh took 186 ms
commit took 1 ms
refresh took 187 ms
commit took 2 ms
refresh took 186 ms
```

**Comment by nirinchev** [comment, marked as answer]
This is because `Refresh()` will block until all subscription handlers have fired. Not only that, but it also forces the otherwise asynchronous changeset calculation to happen synchronously. This is to ensure correctness when notifications are delivered (i.e. they need to see the Realm in a state that is consistent with the changeset indices). I don't know how feasible this is, but the most obvious approach to alleviate this is to move away from calling `Refresh` manually and instead let Realm refresh itself. For this to work though, you'll need to install a synchronization context on your main thread, which may have unintended consequences for other parts of your code (e.g. async continuations being dispatched back on the main thread).

**Reply by peppy** [reply]
> This is because `Refresh()` will block until all subscription handlers have fired. Not only that, but it also forces the otherwise asynchronous changeset calculation to happen synchronously.

Is this implying that the overhead seen here is a real overhead of subscriptions? I have no yet looked into the internals but as I mentioned in the opening post, my assumption was that realm would be smart and efficient with calculating subscription changes, rather than doing a large diff operation over all rows (seems likely what is happening?)

> I don't know how feasible this is, but the most obvious approach to alleviate this is to move away from calling `Refresh` manually and instead let Realm refresh itself. For this to work though, you'll need to install a synchronization context on your main thread, which may have unintended consequences for other parts of your code (e.g. async continuations being dispatched back on the main thread).

We actually already have one installed. I didn't actually realise that we don't need to call `Refresh` ourselves in such cases. Removing the call does 100% alleviate the issue, and we see callbacks firing much less often as expected!

I am still curious about where the changeset overhead comes from in such a simple case as above. If there's any further reading on this please let me know (will probably dig into the source get my head around it regardless).

**Reply by nirinchev** [reply]
The way change calculations happen internally is roughly this:
1. Realm detects that a transaction has been committed bringing it from version 1 to version 2.
2. On a background thread, we look at both versions and for all active subscriptions, we calculate a diff between the collection at version 1 and the collection at version 2.
3. We notify all subscribers on the original thread the Realm was opened at.
4. We wait for 4 to complete, then look for new changes to the Realm file.

When `Refresh` is called manually, it forces all operations to be executed synchronously - notably 2 and 3 will block the caller, so for example if you have expensive detaching going on in a subscription callback, `Refresh` will block until that completes. For point 2., we do try to be smart about change calculation and use the information about which object changed to reduce the calculation overhead, but some things can't be trivially optimized away. For example, if you have a subscription on `Foo` and you insert a `Bar` object, we'll skip change calculation. However, if you have a sort applied, we'll need to sort both collections and then compare the elements in them to calculate the changed indices. I don't have a list of all the heuristics applied during change calculation, but I'm sure there may be areas for improvement.

One side effect of not forcing refresh on every frame is that we can batch changes together. For example, if there are multiple transactions committed while steps 2. and 3. are getting processed for version 2, we'll jump straight to whatever the latest version is, calculating the diff between version 2 and the latest version rather than for each intermediate version.

Finally, refresh is mildly expensive because it involves a managed->native transition and some resource locking in native, thus exacerbating all the issues listed above. Generally speaking, the intention behind it was to use it on background threads where 1) you probably don't have notification listeners and 2) you don't care too much about blocking as delays there should not affect the UX.

If you want to read more about it, you may want to take a look at the `xxx_notifier.cpp` and `xxx_change_xxx.cpp` classes in Core: https://github.com/realm/realm-core/tree/master/src/realm/object-store/impl. I'll admit that the code there goes way over my head, but I'm not a C++ developer, so you may have an easier time understanding what's going on 😅

**Reply by peppy** [reply]
Thanks so much for the detailed explanation.

**Reply by peppy** [reply]
So one (hopefully) final thing: It looks like the same overhead is incurred when opening a new realm context - even if that request is on the main thread with an already open context. I believe either you or someone else on the realm team mentioned previously that opening extra contexts on the same thread will internally share the already open handle, but it seems this is also triggering a refresh/sync implicitly?

![20220120 221833 (Parallels Desktop)](https://user-images.githubusercontent.com/191335/150346252-41aff285-cd9c-4ea4-9a37-6f5b35dc804c.png)

**Reply by nirinchev** [reply]
You're right on both counts - we do get a cached instance and we do refresh it:

https://github.com/realm/realm-dotnet/blob/5beb429be03abec9b7e68c1aa0b19e98701251a6/wrappers/src/shared_realm_cs.cpp#L133-L144

I understand the problem from your end, on the other hand, a lot of users have the expectation that when they call `Realm.GetInstance`, the resulting instance is fully up-to-date with whatever is on disk. For example,

```csharp
await Task.Run(() =>
{
    using var bgRealm = Realm.GetInstance();
    bgRealm.Write(() => { ... });
});

var foregroundRealm = Realm.GetInstance();

// we expect to see whatever we wrote in the background here
```

The caching/refresh part is not necessarily well understood, so we had received bug reports that users are writing some data and not seeing it propagate correctly, which is why we added the forceful refresh. I'll consult with the team and get back to you.

**Reply by peppy** [reply]
I believe we can do something to work around it from our end, so it's not a deal-breaker either way.

**Reply by nirinchev** [reply]
Okay, so spoke with the team and we think that the current behavior better matches user expectations and providing an opt out would be more confusing than helpful. If it turns out more people hit perf issues with it, we'll revisit it.

**Reply by peppy** [reply]
Definitely agree (I checked after this thread and we do rely on said behaviour in a few places).

For anyone curious, we were able to work around the issue by [abstracting realm operations one level further](https://github.com/ppy/osu/blob/8883505eed646f57fdb39c87ad87b8201e99b00b/osu.Game/Database/RealmAccess.cs#L208-L250), where we can funnel all main thread requests into a single realm instance.
